renderdoc/driver/ihv/intel/intel_gl_counters.cpp

```diff
@@ -0,0 +1,318 @@
+    // Used for testing HW is programmed correctly.
+    "TestOa",
+    // Used to plumb raw data from the GL driver to metrics-discovery.
+    "Intel_Raw_Hardware_Counters_Set_0_Query", "Intel_Raw_Pipeline_Statistics_Query"};
+
+IntelGlCounters::IntelGlCounters() : m_passIndex(0)
+{
+}
+
+IntelGlCounters::~IntelGlCounters()
+{
+}
+
+std::vector<GPUCounter> IntelGlCounters::GetPublicCounterIds() const
+{
+  vector<GPUCounter> counters;
+
+  for(const IntelGlCounter &c : m_Counters)
+    counters.push_back(c.desc.counter);
+
+  return counters;
+}
+
+CounterDescription IntelGlCounters::GetCounterDescription(GPUCounter index) const
+{
+  return m_Counters[GPUCounterToCounterIndex(index)].desc;
+}
+
+static CompType glToRdcCounterType(GLuint glDataType)
+{
+  switch(glDataType)
+  {
+    case GL_PERFQUERY_COUNTER_DATA_UINT32_INTEL: return CompType::UInt;
+    case GL_PERFQUERY_COUNTER_DATA_UINT64_INTEL: return CompType::UInt;
+    case GL_PERFQUERY_COUNTER_DATA_FLOAT_INTEL: return CompType::Float;
+    case GL_PERFQUERY_COUNTER_DATA_DOUBLE_INTEL: return CompType::Double;
+    case GL_PERFQUERY_COUNTER_DATA_BOOL32_INTEL: return CompType::UInt;
+    default: RDCERR("Wrong counter data type: %u", glDataType);
+  }
+  return CompType::Typeless;
+}
+
+void IntelGlCounters::addCounter(const IntelGlQuery &query, GLuint counterId)
+{
+  IntelGlCounter counter;
+
+  counter.queryId = query.queryId;
+  counter.desc.counter = (GPUCounter)((int)GPUCounter::FirstIntel + m_Counters.size());
+  counter.desc.category = query.name;
+
+  GLint len = 0;
+  GL.glGetIntegerv(eGL_PERFQUERY_COUNTER_NAME_LENGTH_MAX_INTEL, &len);
+  counter.desc.name.resize(len);
+  GL.glGetIntegerv(eGL_PERFQUERY_COUNTER_DESC_LENGTH_MAX_INTEL, &len);
+  counter.desc.description.resize(len);
+
+  GL.glGetPerfCounterInfoINTEL(
+      query.queryId, counterId, (GLuint)counter.desc.name.size(), &counter.desc.name[0],
+      (GLuint)counter.desc.description.size(), &counter.desc.description[0], &counter.offset,
+      &counter.desc.resultByteWidth, &counter.type, &counter.dataType, NULL);
+
+  if(m_CounterNames.find(counter.desc.name) != m_CounterNames.end())
+    return;
+
+  uint32_t query_hash = strhash(query.name.c_str());
+  uint32_t name_hash = strhash(counter.desc.name.c_str());
+  uint32_t desc_hash = strhash(counter.desc.description.c_str());
+  counter.desc.uuid = Uuid(0x8086, query_hash, name_hash, desc_hash);
+  counter.desc.resultType = glToRdcCounterType(counter.dataType);
+
+  m_Counters.push_back(counter);
+  m_CounterNames[counter.desc.name] = counter;
+}
+
+void IntelGlCounters::addQuery(GLuint queryId)
+{
+  IntelGlQuery query;
+
+  query.queryId = queryId;
+
+  GLint len = 0;
+  GL.glGetIntegerv(eGL_PERFQUERY_QUERY_NAME_LENGTH_MAX_INTEL, &len);
+  query.name.resize(len);
+  GLuint nCounters = 0;
+  GL.glGetPerfQueryInfoINTEL(queryId, (GLuint)query.name.size(), &query.name[0], &query.size,
+                             &nCounters, NULL, NULL);
```

Comment:
These arguments must not be NULL.

Reply:
Do not make comments on PRs that were merged and resolved over a year ago.